### PR TITLE
prometheus-node-exporter-lua: add dawn exporter

### PR DESCRIPTION
**Pull request by PolynomialDivision**
DAWN is a decentralized WiFi Controller.
https://github.com/berlin-open-wireless-lab/DAWN

The node exporter allows to gather statistics about your network.
- Infos about AP (Channel Utilization, Station Count, ...)
- Connected Clients (Signal, Capabilities)

https://github.com/berlin-open-wireless-lab/DAWN/issues/68

### CODE REVIEW

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 1]
```diff
@@ -0,0 +1,49 @@
+local ubus = require "ubus"
+
+local function scrape()
+  local metric_dawn_ap_channel_utilization = metric("dawn_ap_channel_utilization","gauge")
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:4]:
```suggestion
  local metric_dawn_ap_channel_utilization = metric("dawn_ap_channel_utilization_ratio","gauge")
```

If utilization is a percentage it should be in the range 0..1 and suffixed with ratio.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  local metric_dawn_ap_station_count = metric("dawn_ap_station_count","gauge")
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:5]:
```suggestion
  local metric_dawn_ap_station_count = metric("dawn_ap_stations_total","gauge")
```

> Note that an accumulating count has total as a suffix, in addition to the unit if applicable.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+  local metric_dawn_station_signal = metric("dawn_station_signal","gauge")
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:6]:
```suggestion
  local metric_dawn_station_signal = metric("dawn_station_signal_dbm","gauge")
```

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+          local client_ht_support = (client_table['ht'] == true) and 1 or 0
+          local client_vht_support = (client_table['vht'] == true) and 1 or 0
+          local client_signal = client_table['signal'] or -255
+
+          local labels = {
+            ssid = ssid,
+            bssid = ap,
+            mac = client,
+            ht_support = client_ht_support,
+           vht_support = client_vht_support,
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:40]: Indentation is off.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      local ht_support = (ap_table['ht_support'] == true) and 1 or 0
+      local vht_support = (ap_table['vht_support'] == true) and 1 or 0
+
+      local labels = {
+        ssid = ssid,
+        bssid = ap,
+        freq = ap_table['freq'],
+        ht_support = ht_support,
+        vht_support = vht_support,
+        neighbor_report = ap_table['neighbor_report'],
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: What kind of value is this? Labels should be simple strings, nothing more. The neighbor reports should probably be a separate metric.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 1 above)
Comment (PolynomialDivision) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:4]: Channel utilization is a value between 0 and  255. :)

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 5 above)
Comment (PolynomialDivision) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: Neighborreport is just a statis string that won't change. 802.11v is based on it. Maybe I will remove it?

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 1 above)
Comment (mweinelt) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:4]: Ok, so I believe a ratio/percentage would improve the meaning of this metric by scoping it properly.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 5 above)
Comment (mweinelt) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: Oh, so it's more like a boolean, whether the client submits them or is capable of submitting them?

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 1 above)
Comment (PolynomialDivision) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:4]: Yep. :) I will change accordingly, just bumping dawn to latest version.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua (hunk 5 above)
Comment (PolynomialDivision) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: Actually there is information included that allows a client to immidiately roam to some AP (BSSID, Channel, ...)

Comment (mweinelt) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: I guess that can change alot when a client is moving, and labels are not supposted to be short-lived.

Maybe expose those as separate metrics `dawn_station_neighbor_report_dbm` with appropriate labels?

Comment (PolynomialDivision) [reply utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:23]: They don't change. That is static information that APs exchange and give the client on request. :)
Like "Hey AP I want to roam, but I do not want to perform a scan, please give me roaming canidates".
Or what I do: "Hey Client here is a list of APs (as neighborreport) that you should immidiately connect to" :)

**Comment by PolynomialDivision** [comment]
Please merge.

File: utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+        local labels = {
+          ssid = ssid,
+          bssid = ap,
+          freq = ap_table['freq'],
+          ht_support = ht_support,
+          vht_support = vht_support,
+          neighbor_report = ap_table['neighbor_report'],
+        }
+        metric_dawn_ap_channel_utilization_ratio(labels, ap_table['channel_utilization'])
```
Comment (mweinelt) [inline utils/prometheus-node-exporter-lua/files/usr/lib/lua/prometheus-collectors/dawn.lua:27]: Here `ap_table['channel_utilization']` is still 0..255, correct? For ratio it needs to be 0..1.

```suggestion
        metric_dawn_ap_channel_utilization_ratio(labels, ap_table['channel_utilization'] / 255)
```

**Comment by neheb** [comment]
Is this ready to go?

**Comment by PolynomialDivision** [comment]
> Is this ready to go?

From my side everything is fine! :) We changed everything to only export the local node and changed channel utilization to a ratio.

**Comment by champtar** [comment]
@PolynomialDivision you need to bump the version in the Makefile

**Comment by PolynomialDivision** [comment]
> @PolynomialDivision you need to bump the version in the Makefile

Is that correct? Do I have to bump the PKG_RELEASE or the PKG_VERSION? xD Not sure.

**Comment by champtar** [comment]
Bump the PKG_VERSION to today and reset the PKG_RELEASE to 1

**Comment by PolynomialDivision** [comment]
> Bump the PKG_VERSION to today and reset the PKG_RELEASE to 1

done.

**Comment by champtar** [comment]
2020.06.26 you mean

**Comment by PolynomialDivision** [comment]
> 2020.06.26 you mean

sry.

**Comment by champtar** [comment]
Thanks @PolynomialDivision

**Pull request merged**